leakpro/attacks/mia_attacks/abstract_mia.py

```diff
@@ -256,3 +261,117 @@ def run_attack(self:Self) -> Union[AttackResult, List[AttackResult]]:
+        if self.online:
+            for i, (logit, mask, label) in tqdm(enumerate(zip(logits, self.in_indices_mask, self.audit_data._labels))):
+                self.memorization_score[i] = np.mean(logit[mask, label]) - np.mean(logit[~mask, label])
+        else:
+            for i, (logit, target_logit, label) in tqdm(enumerate(zip(logits, target_logits, self.audit_data._labels))):
+                self.memorization_score[i] = np.mean(target_logit[label]) - np.mean(logit[~mask, label])
+
+        self.privacy_score = self._privacy_score()
+
+        mem_mask, privacy_mask = self.adjust_memorization_mask()
+
+        self.skip_indices = (self.skip_indices | mem_mask) | privacy_mask
+
+    def _privacy_score(self:Self) -> np.ndarray:
+        """Run memorization score enhancement.
+
+        Privacy score enhances the attack performance by only inlucing vulnerable data points
+
+        """
+        self.logger.info("Preparing privacy score")
+        if self.signal.__class__.__name__ == "ModelRescaledLogits":
+            logits = self.shadow_models_logits
+            target_logits = self.target_logits
+        else:
+            logits_function = ModelRescaledLogits()
+            logits = np.swapaxes(logits_function(self.shadow_models, self.audit_data), 0, 1)
+            target_logits = np.swapaxes(logits_function([self.target_model], self.audit_data), 0, 1).squeeze()
+
+        self.logger.info("Calculating privacy score")
+        privacy_score = []
+
+        if len(self.shadow_models) < 64:
+            in_std, out_std = np.std(logits[self.in_indices_mask].flatten()), np.std(logits[~self.in_indices_mask].flatten())
+
+        for (logit, target_logit, mask) in tqdm(zip(logits, target_logits, self.in_indices_mask)):
+            in_mean, out_mean = np.mean(logit[mask]), np.mean(logit[~mask])
+
+            if len(self.shadow_models) >= 64:
+                in_std, out_std = np.std(logit[mask]), np.std(logit[mask])
+
+            if self.online:
+                privacy_score.append(np.abs(in_mean-out_mean)/(in_std+out_std+1e-30))
+            else:
+                privacy_score.append(np.abs(target_logit-out_mean)/(2*out_std+1e-30))
+
+        return np.asarray(privacy_score)
+
+    def softmax_logits(self:Self, logits: np.ndarray) -> np.ndarray:
+        """Rescale logits to (0, 1).
+
+        Args:
+        ----
+            logits ( len(dataset) x ... x nb_classes ): Logits to be rescaled.
+
+        """
+        logits = torch.from_numpy(logits)
+        logits = logits - torch.max(logits, dim=-1, keepdim=True).values
+        logits = torch.exp(logits)
+        logits = logits/torch.sum(logits, dim=-1, keepdim=True)
+        return logits.numpy()
+
+    def adjust_memorization_mask(self:Self) -> None:
+        """Adjust thesholds to achieve the desired percentile most vulnerable datapoints."""
+
+        audit_dataset_len = len(self.target_logits)
+        if audit_dataset_len*(1-self.memorization_threshold) < 30:
```

Comment:
maybe using a variable like num_requested_datapoint or something like that instead of a number.